modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/CheckpointPageReplacement.java

```diff
@@ -0,0 +1,136 @@
+
+    private final AtomicBoolean stopGuard = new AtomicBoolean();
+
+    /**
+     * Tries to block the start of the fsync phase at a checkpoint before replacing the page.
+     *
+     * <p>It is expected that if the method returns {@code true}, it will not be invoked a second time with the same page ID.</p>
+     *
+     * @param pageId Page ID for which page replacement is expected to begin.
+     * @return {@code True} if the blocking was successful, {@code false} if the fsync phase is about to begin.
+     * @see #unblock(FullPageId, Throwable)
+     * @see #stopBlocking()
+     */
+    boolean tryBlock(FullPageId pageId) {
+        if (!busyLock.enterBusy()) {
+            return false;
+        }
+
+        try {
+            boolean added = pageIds.add(pageId);
+
+            assert added : "Page is already in the process of being replaced: " + pageId;
+
+            return true;
+        } finally {
+            busyLock.leaveBusy();
+        }
+    }
+
+    /**
+     * Unblocks the start of the fsync phase at a checkpoint after the page replacement is completed.
+     *
+     * <p>It is expected that the method will be invoked once and after the {@link #tryBlock(FullPageId)} returns {@code true} for same
+     * page ID.</p>
+     *
+     * <p>The fsync phase will only be started after page replacement has been completed for all pages for which
+     * {@link #tryBlock(FullPageId)} was invoked before {@link #stopBlocking()} was invoked, or no page replacement occurred at all.</p>
+     *
+     * <p>If any error occurred during page replacement, then the future from {@link #stopBlocking()} will be completed with the first
+     * error.</p>
+     *
+     * @param pageId Page ID for which the page replacement has ended.
+     * @param error Error on page replacement, {@code null} if missing.
+     * @see #tryBlock(FullPageId)
+     * @see #stopBlocking()
+     */
+    void unblock(FullPageId pageId, @Nullable Throwable error) {
+        boolean removed = pageIds.remove(pageId);
+
+        assert removed : "Replacement for the page either did not start or ended: " + pageId;
+
+        if (error != null) {
+            stopBlockingFuture.completeExceptionally(error);
+
+            return;
+        }
+
+        if (!busyLock.enterBusy()) {
+            if (pageIds.isEmpty()) {
+                stopBlockingFuture.complete(null);
+            }
+        } else {
+            busyLock.leaveBusy();
+        }
+    }
+
+    /**
+     * Stops new blocks before the fsync phase starts at a checkpoint.
+     *
+     * @return Future that will be completed successfully if all blocks are completed before the current method is invoked, either if there
+     *      were none, or with an error from the first unlock.
+     * @see #tryBlock(FullPageId)
+     * @see #unblock(FullPageId, Throwable)
+     */
+    CompletableFuture<Void> stopBlocking() {
+        if (stopGuard.compareAndSet(false, true)) {
```

Comment:
I wonder if we actually try stopping it twice. Why do we need this protection?

Reply:
They shouldn't call twice, I added it just in case, are you suggesting to get rid of it?

Reply:
I don't know, it's just a couple of lines of code, so if it stays here I won't be mad